health/metrics/plugin_impl_metrics.go

```diff
@@ -0,0 +1,234 @@
+type Plugin struct {
+	Deps
+
+	customProbe bool
+}
+
+// Deps lists dependencies of the Prometheus plugin.
+type Deps struct {
+	local.PluginLogDeps                               // inject
+	HTTP                *rest.Plugin                  // inject (optional)
+	StatusCheck         statuscheck.AgentStatusReader // inject
+}
+
+// Init may create a new (custom) instance of HTTP if the injected instance uses
+// different HTTP port than requested.
+func (p *Plugin) Init() (err error) {
+	// Start Init() and AfterInit() for new Prometheus in case the port is different
+	// from agent http.
+	if p.HTTP.HTTPport != httpPort {
+		childPlugNameHTTP := p.String() + "-HTTP"
+		p.HTTP = &rest.Plugin{
+			Deps: rest.Deps{
+				Log:        logging.ForPlugin(childPlugNameHTTP, p.Log),
+				PluginName: core.PluginName(childPlugNameHTTP),
+				HTTPport:   httpPort,
+			},
+		}
+		err := p.HTTP.Init()
+		if err != nil {
+			return err
+		}
+		err = p.HTTP.AfterInit()
+		if err != nil {
+			return err
+		}
+
+		p.customProbe = true
+	}
+
+	p.RegisterGauge(
+		namespace,
+		subsystem,
+		serviceHealthName,
+		serviceHealthHelp,
+		prometheus.Labels{serviceLabel: agentName},
+		p.getServiceHealth,
+	)
+
+	agentStatus := p.StatusCheck.GetAgentStatus()
+	p.RegisterGauge(
+		namespace,
+		subsystem,
+		serviceInfoName,
+		serviceInfoHelp,
+		prometheus.Labels{
+			serviceLabel:      agentName,
+			buildVersionLabel: agentStatus.BuildVersion,
+			buildDateLabel:    agentStatus.BuildDate},
+		func() float64 { return 1 },
+	)
+
+	return nil
+}
+
+// AfterInit registers HTTP handlers.
+func (p *Plugin) AfterInit() error {
+	if p.HTTP != nil {
+		if p.StatusCheck != nil {
+			p.Log.Infof("Starting Prometheus metrics handlers on port %v", p.HTTP.HTTPport)
+			p.HTTP.RegisterHTTPHandler(DefaultMetricsPath, p.metricsHandler, "GET")
+		} else {
+			p.Log.Info("Unable to register Prometheus metrics handlers, StatusCheck is nil")
+		}
+	} else {
+		p.Log.Info("Unable to register Prometheus metrics handlers, HTTP is nil")
+	}
+
+	return nil
+}
+
+// Close shutdowns HTTP if a custom instance was created in Init().
+func (p *Plugin) Close() error {
+	if p.customProbe {
+		_, err := safeclose.CloseAll(p.HTTP)
+		return err
+	}
+
+	return nil
+}
+
+// metricsHandler handles Prometheus metrics collection.
+func (p *Plugin) metricsHandler(formatter *render.Render) http.HandlerFunc {
+	return promhttp.Handler().ServeHTTP
+}
+
+func (p *Plugin) getServiceHealth() float64 {
+	agentStatus := p.StatusCheck.GetAgentStatus()
+	// Adapt Ligato status code for now.
+	// TODO: Consolidate with that from the "Common Container Telemetry" proposal.
+	health := float64(agentStatus.State)
+	log.StandardLogger().Infof("getServiceHealth(): %f", health)
+	return health
+}
+
+// RegisterGauge registers custom gauge with specific valueFunc to report status when invoked.
+func (p *Plugin) RegisterGauge(namespace string, subsystem string, name string, help string,
```

Comment:
This method seems to be similar to statuscheck.PluginStatusWriter.Register.

What about not exposing new register function for doing simmilar thing.

I was expecting that:
- statuscheck is just in memory "registry" with abillity to poll status on previously registered plugins
- this plugin is just another format for probes

Suggestion:
- let's add parameters (subsystem ...) to existing statuscheck.PluginStatusWriter.Register
- let's expose pluginstatusmap in statuscheck
- let's consume pluginstatusmap in this new plugin and basically just integrate with prometheus

Reply:
For now, I will change it to register() for reference -- I think one would need the info there when designing/changing code for the above.

Reply:
If you were to do the above enhancements, please note that Prometheus metrics registration may not always go through
prometheus.DefaultRegisterer.Register()
I have a user story to explore the option to separate health metrics from the built-in runtime metrics.  This implies the health metrics need to be registered/implemented in different way.